packages/firestore/src/platform_node/grpc_connection.ts

```diff
@@ -273,23 +264,19 @@ export class GrpcConnection implements Connection {
       close(new FirestoreError(code, grpcError.message));
     });
 
-    grpcStream.on('status', (status: grpc.StatusObject) => {
-      // TODO(mikelehen): I *believe* this assert is safe and we can just remove
-      // the 'status' event if we don't see the assert getting hit for a while.
-      assert(
-        closed,
-        `status event received before "end" or "error". ` +
-          `code: ${status.code} details: ${status.details}`
-      );
-    });
-
     log.debug(LOG_TAG, 'Opening GRPC stream');
-    // TODO(dimond): Since grpc has no explicit open status (or does it?) we
-    // simulate an onOpen in the next loop after the stream had it's listeners
-    // registered
-    setTimeout(() => {
-      stream.callOnOpen();
-    }, 0);
+
+    // TODO(mikelehen): We will call waitForReady() on each stream attempt,
+    // but waitForReady() won't call our callback until the channel is
+    // successfully established (note that the callback is not called for most
+    // stream errors). So we'll accumulate more and more outstanding wait calls
+    // over time while the network is down, manifesting as a minor memory leak.
```

Comment:
I can solve this TODO if you want but I think it's going to be a bit ugly so wanted to check with you first to see if we really care...  We basically need a singleton (per stub) ChannelWaiter object that you can subscribe to/from (and make sure that streams unsubscribe when they get an error).  

And to be fully leak-proof, it may need to use the raw connectivityState API rather than waitForReady(), but that isn't exposed in grpc's .d.ts file so we'd be in un-typed territory.

Reply:
How is this any different than the new stream we're creating per invocation above?

Couldn't we make this another callback on the stream bridge?

Reply:
`grpcStream` is short-lived.  We create it on each retry and should throw it away once it fails.  Then we create a new one for the next retry.  No leak.

`client` is long-lived (it's the stub that we create per-token and lives at least for the life of the token, possibly longer).  Additionally our `client.waitForReady()` callbacks are long-lived (when the network is down) because they're not called until the client/channel successfully connects (potentially after many retry attempts).

So after 10 retries, we'll only have a single `grpcStream` of interest, but we'll have queued 10 `waitForReady()` callbacks.  Once we actually connect, all 10 will be called by grpc but all but the most recent one will be no-ops because those streams are closed / forgotten.

Note I haven't done any deep analysis of this code to convince myself we're not leaking lots of other things.  But I'm pretty certain we can easily build up an unbounded number of `waitForReady()` callbacks...  That said our max backoff is 60 seconds. With 2 streams, retrying once a minute, that's `2 * 24 * 60 = 2880` callbacks leaked per day of being offline and retrying.  I doubt that's a noticeable memory or CPU impact, so I'm semi-inclined not to care. :-)